tfx_bsl/beam/run_inference.py

```diff
@@ -140,76 +249,135 @@ def RunInferenceImpl(  # pylint: disable=invalid-name
+def _Classify(
+    pcoll: beam.pvalue.PCollection,
+    fixed_inference_spec_type: model_spec_pb2.InferenceSpecType = None
+): # pylint: disable=invalid-name
   """Performs classify PTransform."""
-  if _using_in_process_inference(inference_spec_type):
-    return (pcoll
-            | 'Classify' >> beam.ParDo(
-                _BatchClassifyDoFn(inference_spec_type, shared.Shared()))
-            | 'BuildPredictionLogForClassifications' >> beam.ParDo(
-                _BuildPredictionLogForClassificationsDoFn()))
+  raw_predictions = None
+
+  if fixed_inference_spec_type is None:
+    tagged = pcoll | 'Tag inference type' >> _TagUsingInProcessInference()
+    tagged['remote'] | 'NotImplemented' >> _NotImplementedTransform()
+    raw_predictions = (
+      tagged['local']
+      | 'Classify' >> beam.ParDo(_BatchClassifyDoFn(shared.Shared())))
   else:
-    raise NotImplementedError
+    if _using_in_process_inference(fixed_inference_spec_type):
+      raw_predictions = (
+        pcoll
+        | 'Classify' >> beam.ParDo(_BatchClassifyDoFn(shared.Shared(),
+            fixed_inference_spec_type=fixed_inference_spec_type)))
+    else:
+      raise NotImplementedError
+
+  return (raw_predictions
+          | 'BuildPredictionLogForClassifications' >> beam.ParDo(
+              _BuildPredictionLogForClassificationsDoFn()))
 
 
 @beam.ptransform_fn
-@beam.typehints.with_input_types(Union[tf.train.Example,
-                                       tf.train.SequenceExample])
+@beam.typehints.with_input_types(_QueryBatchType)
 @beam.typehints.with_output_types(prediction_log_pb2.PredictionLog)
-def _Regress(pcoll: beam.pvalue.PCollection,  # pylint: disable=invalid-name
-             inference_spec_type: model_spec_pb2.InferenceSpecType):
+def _Regress(
+    pcoll: beam.pvalue.PCollection,
+    fixed_inference_spec_type: model_spec_pb2.InferenceSpecType = None
+): # pylint: disable=invalid-name
   """Performs regress PTransform."""
-  if _using_in_process_inference(inference_spec_type):
-    return (pcoll
-            | 'Regress' >> beam.ParDo(
-                _BatchRegressDoFn(inference_spec_type, shared.Shared()))
-            | 'BuildPredictionLogForRegressions' >> beam.ParDo(
-                _BuildPredictionLogForRegressionsDoFn()))
+  raw_predictions = None
+
+  if fixed_inference_spec_type is None:
+    tagged = pcoll | 'Tag inference type' >> _TagUsingInProcessInference()
+    tagged['remote'] | 'NotImplemented' >> _NotImplementedTransform()
+    raw_predictions = (
+      tagged['local']
+      | 'Regress' >> beam.ParDo(_BatchRegressDoFn(shared.Shared())))
   else:
-    raise NotImplementedError
+    if _using_in_process_inference(fixed_inference_spec_type):
+      raw_predictions = (
+        pcoll
+        | 'Regress' >> beam.ParDo(_BatchRegressDoFn(shared.Shared(),
+            fixed_inference_spec_type=fixed_inference_spec_type)))
+    else:
+      raise NotImplementedError
```

Comment:
Seems this part is repeated several times, can this be extracted to a function?

Reply:
Refactored this section into a single operation constructor: https://github.com/tensorflow/tfx-bsl/blob/a88ecc19164b76362ba16e20d8b59a02a765a7ad/tfx_bsl/beam/run_inference.py#L940-L1020